delphi-frontend/src/main/java/au/com/integradev/delphi/symbol/resolve/TypeInferrer.java

```diff
@@ -48,8 +50,15 @@ public Type infer(Typed typed) {
 
       if (arrayConstructor instanceof ArrayConstructorNode) {
         type = inferArrayConstructor((ArrayConstructorNode) arrayConstructor);
-      } else if (ExpressionNodeUtils.isIntegerLiteral(expression) && type.size() <= 4) {
-        type = typeFactory.getIntrinsic(IntrinsicType.INTEGER);
+      } else {
+        IntegerLiteralNode literal = ExpressionNodeUtils.unwrapInteger(expression);
+        if (literal != null) {
+          IntegerType integer = (IntegerType) typeFactory.getIntrinsic(IntrinsicType.INTEGER);
+          if (integer.min().compareTo(literal.getValue()) <= 0
+              && integer.max().compareTo(literal.getValue()) >= 0) {
+            type = typeFactory.getIntrinsic(IntrinsicType.INTEGER);
```

Comment:
Should this return an inferred type if it's out of these ranges? e.g. `Cardinal` or `Int64`? I'm not sure how Delphi's inference works here.

Reply:
The weird part of the inference process is just widening anything within the `Integer` range to `Integer`.
Otherwise, we take the type of the expression based on the literal value.
Based on that, the possible integer types are `Integer` `Cardinal`, `Int64`, and `UInt64`. This is reflected in the unit tests.